src/linker/Linker.Dataflow/FlowAnnotations.cs

```diff
@@ -0,0 +1,340 @@
+﻿using Mono.Cecil;
+using Mono.Cecil.Cil;
+using System;
+using System.Collections.Generic;
+
+namespace Mono.Linker.Dataflow
+{
+	// TODO: enforce virtual methods have consistent annotations
+
+	class FlowAnnotations
+	{
+		readonly IFlowAnnotationSource _source;
+		readonly Dictionary<TypeDefinition, TypeAnnotations> _annotations = new Dictionary<TypeDefinition, TypeAnnotations>();
+
+		public FlowAnnotations (IFlowAnnotationSource annotationSource, LinkContext context)
+		{
+			_source = annotationSource;
+		}
+
+		public bool RequiresDataFlowAnalysis (MethodDefinition method)
+		{
+			return GetAnnotations (method.DeclaringType).TryGetAnnotation (method, out _);
+		}
+
+		public bool RequiresDataFlowAnalysis (FieldDefinition field)
+		{
+			return GetAnnotations (field.DeclaringType).TryGetAnnotation (field, out _);
+		}
+
+		/// <summary>
+		/// Retrieves the annotations for the given parameter.
+		/// </summary>
+		/// <param name="parameterIndex">Parameter index in the IL sense. Parameter 0 on instance methods is `this`.</param>
+		/// <returns></returns>
+		public DynamicallyAccessedMemberKinds GetParameterAnnotation (MethodDefinition method, int parameterIndex)
+		{
+			if (GetAnnotations (method.DeclaringType).TryGetAnnotation (method, out var annotation)) {
+				return annotation.ParameterAnnotations [parameterIndex];
+			}
+
+			return 0;
+		}
+
+		public DynamicallyAccessedMemberKinds GetReturnParameterAnnotation (MethodDefinition method)
+		{
+			if (GetAnnotations (method.DeclaringType).TryGetAnnotation (method, out var annotation)) {
+				return annotation.ReturnParameterAnnotation;
+			}
+
+			return 0;
+		}
+
+		public DynamicallyAccessedMemberKinds GetFieldAnnotation (FieldDefinition field)
+		{
+			if (GetAnnotations (field.DeclaringType).TryGetAnnotation (field, out var annotation)) {
+				return annotation.Annotation;
+			}
+
+			return 0;
+		}
+
+		private TypeAnnotations GetAnnotations(TypeDefinition type)
+		{
+			if (!_annotations.TryGetValue(type, out TypeAnnotations value)) {
+				value = BuildTypeAnnotations (type);
+				_annotations.Add (type, value);
+			}
```

Comment:
This means we will end up with a `TypeAnnotations` instance (along with several empty arrays on it) for every type in the system. Given that there will be proportionally tiny number of interesting types that seems wasteful. Can we store nulls in the dictionary for the cases where the type is not interesting (no annotations anywhere on it)? That should help a lot.

Reply:
I see it's a struct sorry - my bad. I guess it's OK then... it's still larger than one pointer per type, but not that bad.